### PR TITLE
MH-12704 Captions support for paella

### PR DESCRIPTION
**Pull request by miesgre**
Captions support 
* Captions must be flavored as 'captions/timedtext'.
* The caption element must be tagged with "lang:XX" (lang:en, lang:es, ...) to specify the captions language.
* The captions format is figured out by the file extensions
Available formats:
  - vtt
  - dfxp

### CODE REVIEW

**Comment by miesgre** [comment]
@karendolan, @rute-santos Can you test this PR?

**Review by staubesv — CHANGES_REQUESTED** [review]
Hi @miesgre, @karendolan & @rute-santos: Just started to test this, but ran into the following problem:

Using the flexible asset upload facility, I can configure Opencast to support the upload of captions. While it is possible to configure the type (catalog or attachment) and the flavor (e.g. captions/timedtext) of uploads, it is not possible to configure tags (missing: lang:*).

@miesgre This implies that the current upload capabilities of Opencast cannot be used for captions upload (which is part of what they are supposed to be useful for) with this code.

btw: Why are you actually looking at tags for DFXP captions?
The DFXP example on https://www.speechpad.com/captions/dfxp suggests that DFXP XMLs can contain both information about the language and also multiple languages in a single DFXP file.

WebVTT seems less sophisticated - I couldn't find a way how a WebVTT would self-describe the language of the captions.

@karendolan, @rute-santos: An obvious way to improve the situation at least for WebVTT would be to extend the flexible upload facility configuration options so that a list of tags could be specified for each configured upload. That way, uploading captions for WebVTT would work. What do you think?

**Comment by rute-santos** [comment]
Hi @staubesv,

The idea to use tags was exaclty because of webvtt. Alternatively, Paella could try to extract the language information from the file (dfxp) and fall back to the tag if the language was not found. Would that be a better solution?

Regarding the captions upload, why can't you tag the uploaded element in the workflow since they have a known flavor?

@miesgre We've just come back from vacation and will test this in the next few days.

**Comment by staubesv** [comment]
Oh.... totally forgot that the upload will case a workflow to be started.... ok, I can set tags ;-)

Will continue testing...

**Comment by miesgre** [comment]
Hi @staubesv, I decided to use the tag "lang:XX" in all formats because in this way all formats works in the same way. Not in dfxp you need to put the lang inside the captios, but in vtt, srt or whatever you need to put the tag "lang:XX".

But, I can modify to extract the lang from the dfxp if desired.
For now, paella supports only one language per dfxp file.

**Comment by miesgre** [comment]
@staubesv, @karendolan I was thinking that the language can be extracted also from the filename without using the tag "lang:XX".
For Example:
- captions.en.vtt
English lang
- captions.es.vtt
Spanish lang

What do you think?

**Comment by rute-santos** [comment]
Hi @miesgre, I think that the use of tags is more flexible. It's easier to tag an element in a media package than making sure the file has the correct name (I actually don't know how to do that).

**Comment by rute-santos** [comment]
Hi @miesgre, the vtt caption worked fine, the dfxp didn't. I believe it's because it's a catalog and has the 'xml' extension and the extension is used by the player to identify the format. 
Can we assume all '.xml' caption catalogs are dfxp? Other options would be: 1) Paella could read the file to figure out the format, 2) The caption converter service could generate the file with a .dfxp extension and as an attachment (but this is not backwards compatible so I don't know if it could cause problems...). @karendolan, @staubesv, ideas?
I am inclined to go with option 2 unless there are objections. The file extension and mime-type for the dfxp caption are lost when the element is added as a catalog to the media package (the extension becomes 'xml' and the mime-type 'text/xml') so it makes it difficult for the player to figure out the correct format.

**Comment by staubesv** [comment]
@miesgre 

> But, I can modify to extract the lang from the dfxp if desired.
> For now, paella supports only one language per dfxp file.

I'm ok with paella only supporting one language der dfxp file for now. Extending this could be done later if required.

> @rute-santos 
> Can we assume all '.xml' caption catalogs are dfxp? Other options would be: 1) Paella could read the file to figure out the format, 2) The caption converter service could generate the file with a .dfxp extension and as an attachment (but this is not backwards compatible so I don't know if it could cause problems...). @karendolan, @staubesv, ideas?

\1) seems not technically sound as .xml simply doesn't imply dfxp. \2) has similar issues as .dfxp is not the only file extensions DFXP files in general can have (according to Wikipedia, .xml and .ttml seem also valid). Don't forget here that the user can manually upload captions, too. Anf if .xml and .ttml files are generated by external captioning tool, it is desirable that a user can upload those without needing to change the file extension because of (unobvious) "workarounds".

It might be worth to use a separate flavor to locate dfxp catalogs. That would avoid relying on file extensions or mime-types. E.g. captions/dfxp or something like that. 

@miesgre If the dfxp catalog would be located based on a dfxp-specific flavor, it would then make more sense to omit that lang:xx tags (for dfxp only) and parse the language from the catalog itself.
This also avoids unobivous support cases: Too bad if a user uploads a DXFP with "internal language" set to Spanish but tagged as German ;-)

Anyway... I don't have a strong opinion about this...

**Comment by miesgre** [comment]
@staubesv, @rute-santos  What about this?

Captions are tagged as `captions/<format>` (ex: captions/dfxp, caption/vtt, captions/srt, etc...)
So the captions format will be the subflavor. Makes sense?

About the language. Paella can try to figure out using these rules:

1. If format is dfxp, read the language from the file
2. Read the lang from a tag `lang:XX` if tag exists
3. Try to figure out the lang from file extension (file.es.vtt)

If all above fails paella can continue showing captions but we can mark it as `unknown language`

I think captions should be attachments, but for backwards compatibility paella can try to load a `captions/timedtext` catalogs and asume that they are dfxp

**Comment by rute-santos** [comment]
@miesgre You meant flavored as `captions/<format>`, not tagged, correct? I think this is a very good idea. I will update the caption service (used by the transcription service) to flavor the element caption/dfxp, caption/vtt, etc and add them as attachments and also to add the lang:XX tag. @staubesv I believe this solution addresses your concerns?

**Comment by staubesv** [comment]
> @miesgre You meant flavored as captions/<format>, not tagged, correct? I think this is a very good idea. I will update the caption service (used by the transcription service) to flavor the element caption/dfxp, caption/vtt, etc and add them as attachments and also to add the lang:XX tag. @staubesv I believe this solution addresses your concerns?

Fine for me.

**Comment by miesgre** [comment]
@rute-santos Yes, flavored. Sorry for the misunderstanding 📦 

I was thinking... what about adding an optional `+lang` to the subflavour?  So, we can avoid the use of tags.

ex: `captions/dfxp`, `captions/dfxp+en`, `captions/vtt+es`, ...

**Comment by rute-santos** [comment]
@miesgre No objections!

**Comment by rute-santos** [comment]
The transcription service updates are in PR #109.

**Comment by miesgre** [comment]
@rute-santos @staubesv  I updated the PR.

Paella searches the captions in the attachments flavored as `captions/<format>` (ex `captions/dfxp`, `captions/vtt`,...)

The lang can be added to the subfravour `+<lang>` (ex: `captions/vtt+en`) or can be added as a tag `lang:en`
If lang language is not detected, captions will be marked as `unknow language`, but they will be visible.

For backwards compatibility paella loads the `captions/timedtext` catalogs and asume that they are dfxp.

**Comment by rute-santos** [comment]
One more question: should we use 'vtt' to 'webvtt'? Or accept both? It doesn't matter to me.

**Comment by rute-santos** [comment]
Hi @miesgre, I tested with vtt, dfxp as attachments, and dfxp as captions/timedtext. When there's no language information neither in the flavor nor tag, the captions are not displayed in the player. Is this the intended behavior? If the language info is there, everything works fine.

**Comment by miesgre** [comment]
@rute-santos, Is the captions button visible? 

Note that paella only autoselect captions if captions language is equal to your browser language. Otherwise you need to select the desired captions in the combo box.

Please check that the captions button is visible. If visible, push it and check the combo box to select the captions language masked as "unknown".

Maybe, for better UI experience, paella can autoselect the caption if only one is available. I will implement this for next paella release :P

**Comment by rute-santos** [comment]
@miesgre, you are correct. The captions are there, I just needed to select the 'unknown language'. So everything seems to be working fine!

**Comment by miesgre** [comment]
@staubesv @karendolan Do you agree the last changes? Can this be merged? or are more changes needed?

**Comment by karendolan** [comment]
@miesgre It looks good to me :thumbsup:

File: etc/listproviders/event.upload.asset.options.properties [hunk 1]
```diff
@@ -30,8 +30,8 @@ list.name=eventUploadAssetOptions
 # Attachments and catalogs upload options are for new and existing events.
 # Only one file can be uploaded for each of these options: the uploaded file replaces existing elements of the same type and flavor in the mediapackage.
 # EVENTS.EVENTS.NEW.UPLOAD_ASSET.OPTION.CLASS_HANDOUT_NOTES={"id": "attachment_class_handout_notes", "type": "attachment", "flavorType": "attachment", "flavorSubType": "notes", "displayOrder":1}
-# EVENTS.EVENTS.NEW.UPLOAD_ASSET.OPTION.CAPTIONS_DFXP={"id":"catalog_captions_dfxp", "type": "catalog", "flavorType": "captions", "flavorSubType": "timedtext", "displayOrder":2}
-# EVENTS.EVENTS.NEW.UPLOAD_ASSET.OPTION.CAPTIONS_WEBVTT={"id":"attachment_captions_webvtt", "type": "attachment", "flavorType": "text", "flavorSubType": "webvtt", "displayOrder":3}
+# EVENTS.EVENTS.NEW.UPLOAD_ASSET.OPTION.CAPTIONS_DFXP={"id":"catalog_captions_dfxp", "type": "attachment", "flavorType": "captions", "flavorSubType": "dfxp", "displayOrder":2}
```
Comment (staubesv) [inline etc/listproviders/event.upload.asset.options.properties]: @karendolan Interestingly, this change introduces the following problem:

Reproduction:
1. Uncomment line 33
2. Add an event (add a dfxp catalog in the Add Event wizard)

This results in:

`2018-03-20 10:45:26,593 | WARN  | (IndexServiceImpl:1228) - Unable to process asset metadata [{"flavorType":"presenter","multiple":false,"displayOrder":13,"id":"track_presenter","type":"track","title":"EVENTS.EVENTS.NEW.SOURCE.UPLOAD.NON_SEGMENTABLE","flavorSubType":"source"},{"flavorType":"captions","displayOrder":2,"id":"catalog_captions_dfxp","type":"attachment","title":"EVENTS.EVENTS.NEW.UPLOAD_ASSET.OPTION.CAPTIONS_DFXP","flavorSubType":"dfxp"},{"flavorType":"presentation","multiple":false,"displayOrder":14,"id":"track_presentation","type":"track","title":"EVENTS.EVENTS.NEW.SOURCE.UPLOAD.SEGMENTABLE","flavorSubType":"source"},{"flavorType":"text","displayOrder":3,"id":"attachment_captions_webvtt","type":"attachment","title":"EVENTS.EVENTS.NEW.UPLOAD_ASSET.OPTION.CAPTIONS_WEBVTT","flavorSubType":"vtt"}]
`

I have verified that changed catalog_captions_dfxp to attachment_captions_dfxp fixes the problem... OMG!

@miesgre: Seems you need to fix this ;-)

@karendolan Do you know why this happens?

Comment (karendolan) [reply etc/listproviders/event.upload.asset.options.properties]: The id prefix needs to match the type value. It’s a shortcut to sorting the form items on the server side.

Comment (miesgre) [reply etc/listproviders/event.upload.asset.options.properties]: @staubesv Done!

**Review by staubesv — APPROVED** [review]

**Pull request merged**